source/fundamentals/crud/read-operations/cursor.txt

```diff
@@ -1,8 +1,266 @@
+      cursor, err := coll.Find(context.TODO(), bson.D{}, opts)
+      if err != nil {
+         panic(err)
+      }
+
+Errors
+~~~~~~
+
+To retrieve the last error the cursor encountered, use the ``Err()``
+function. The error is ``nil`` if no errors occurred.
+
+.. literalinclude:: /includes/fundamentals/code-snippets/CRUD/cursor.go
+   :language: go
+   :dedent:
+   :start-after: begin err
+   :end-before: end err
+
+.. _cursor-iteration-golang:
+
+Usage Patterns
+--------------
+
+A cursor allows you to access query results one document at a time,
+abstracting away network and caching logic. 
+
+.. note::
+   
+   If you previously used the cursor to iterate some documents, the
+   driver excludes those documents when you iterate again. 
+
+Functional Iteration
+~~~~~~~~~~~~~~~~~~~~
+
+Functional iteration functions make your cursor non-tailable, which
+means the following occurs:
+
+- The cursor automatically closes after the ``MongoClient`` exhausts it.
+- The cursor stops retrieving documents even if you insert them.
+
+All
+```
+
+To iterate through all results, use the ``All()`` function. 
+
+This function unmarshals all the documents stored in a cursor into an
+array at the same time. 
+
+.. literalinclude:: /includes/fundamentals/code-snippets/CRUD/cursor.go
+   :language: go
+   :dedent:
+   :start-after: begin cursor all
+   :end-before: end cursor all
+
+Manual Iteration
+~~~~~~~~~~~~~~~~
+
+Manual iteration functions make your cursor tailable, which means the
+following occurs:
+
+- The cursor remains open after the ``MongoClient`` exhausts it.
+- The cursor continues to retrieve documents as you insert them.
+
+Next
+````
+
+To iterate through each result or to purposely wait for the next
+available document, use the ``Next()`` function. 
+
+.. tip::
+
+   While going through each result, you can further specify how you want
+   the documents printed though the :ref:`find operations options
+   <retrieve-options>`.
+
+This function blocks the current goroutine until a document is
+available, an error occurs, or the context expires. It returns ``true``
+if no errors occur and the ``MongoClient`` didn't exhaust the cursor.
+Otherwise, the function returns ``false``.
+   
+.. literalinclude:: /includes/fundamentals/code-snippets/CRUD/cursor.go
+   :language: go
+   :dedent:
+   :start-after: begin cursor next
+   :end-before: end cursor next
+
+Attempt Next
+````````````
+
+Alternatively, you can use the ``TryNext()`` function.
+
+This function attempts to retrieve the next document in a cursor. It's
+non-blocking, which means it returns ``false`` if the ``MongoClient``
+exhausts the cursor, an error occurs, the next document isn't available
+or the context expires. Otherwise, the function returns ``true``.
+
+.. literalinclude:: /includes/fundamentals/code-snippets/CRUD/cursor.go
+   :language: go
+   :dedent:
+   :start-after: begin cursor try next
+   :end-before: end cursor try next
+
```

Comment:
Suggestion: Could "tailable" be defined or omitted?
Suggestion: Could "Manual Iteration" be given a Task-Oriented title?
Issue: Could "iterate through each" (behavior of `All` method) be changed to "gets the next"?
Issue: I believe `Next` is a method
Suggestion: Could a detailed discussion of return values be saved for API documentation?

This suggestion is a rough draft, but I think it proposes potential solutions to some of the items listed above. 

```suggestion
Iterate Through Your Documents Individually
---------------------------------------------

To get the next result document from your cursor, use the ``Next()`` method.
The next method fetches your result document from Mongo
   
.. literalinclude:: /includes/fundamentals/code-snippets/CRUD/cursor.go
   :language: go
   :dedent:
   :start-after: begin cursor next
   :end-before: end cursor next

.. important:: Blocking

   This function blocks the current goroutine until a document is
   available, an error occurs, or the context expires. 

Retrieve the Next Document if it is currently in the Cursor
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

If you only want to return the next document if it is already in your Cursor,
use the ``TryNext()`` function:

.. literalinclude:: /includes/fundamentals/code-snippets/CRUD/cursor.go
   :language: go
   :dedent:
   :start-after: begin cursor try next
   :end-before: end cursor try next

```

Reply:
For each section, I didn't include a "The following code snippet ..." description because I felt that it would be reiterating what was mentioned right before it.